### PR TITLE
[SPMD] Support SymInt with non-op call_function nodes

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99504
* #99501
* __->__ #99420
* #99417
* #99363
* #99332
* #99331

### CODE REVIEW

File: torch/distributed/_spmd/distribute.py [hunk 1]
```diff
@@ -744,6 +744,15 @@ def _convert_to_distributed(
...
             args = tree_map(partial(_remap_arg, node_to_obj), node.args)
             kwargs = tree_map(partial(_remap_arg, node_to_obj), node.kwargs)
+
+            # FIXME: non-op functions always use local value of SymInt for now
+            args = tree_map(
+                lambda a: a.local_value if isinstance(a, DSymInt) else a, args
+            )
+            kwargs = tree_map(
+                lambda a: a.local_value if isinstance(a, DSymInt) else a, kwargs
+            )
```
Comment (fegin) [inline torch/distributed/_spmd/distribute.py:754]: For math operations, shouldn't make the result DSymInt so that the sharding is correctly propagate?

Comment (mrshenli) [reply torch/distributed/_spmd/distribute.py:744]: good point! Refactored DSymInt to support DSymInt propagation.  :)

File: torch/distributed/_spmd/distribute.py [hunk 2]
```diff
@@ -358,24 +351,21 @@ def _get_dtensor_dispatch_graph(
...
                 )
+                return None
             else:
-                # If an operator consumes SymInt sizes on a sharded dimension, we
-                # override with callables in SYM_INT_CONSUMERS or FACTORY_OPS to
-                # create DTensor activations.
-                raise NotImplementedError(
-                    f"{op_overload} consumes SymInt args from a sharded dimension, "
-                    "but SPMD expansion does not support this use case."
+                logger.warn(
```
Comment (mrshenli) [inline torch/distributed/_spmd/distribute.py:365]: Adding DSymInt support for all ops (e.g., `aten.add.Scalar`, `aten.add.Tensor`, `aten.add.int`, etc.) seems to be unnecessary. It seems to be sufficient to only special-case view and factory ops. All other ops can consume the local_value of DSymInt.

File: torch/distributed/_spmd/distribute.py [hunk 3]
```diff
@@ -689,6 +684,9 @@ def _convert_to_distributed(
     """
     global logger
     logger = get_logger("spmd_exp")
+    operators = {
```
Comment (fegin) [inline torch/distributed/_spmd/distribute.py:687]: nit: can be simplified with `operators = operator.__all__`

File: torch/distributed/_spmd/distribute.py [hunk 4]
```diff
@@ -740,11 +738,47 @@ def _convert_to_distributed(
...
+            dsymints = list(
+                filter(lambda a: isinstance(a, DSymInt), args + tuple(kwargs.values()))
+            )
+
+            if node.target in operators and len(dsymints) > 0:
+                assert all(
+                    dsymints[0].mesh == d.mesh for d in dsymints
+                ), "all DSymInts must have the same mesh. "
+
+                local_args = tree_map(
```
Comment (fegin) [inline torch/distributed/_spmd/distribute.py:754]: nit: these mapping can be simplified with `tree_map_only`
```
local_args = tree_map_only(DSymInt, lambda a: a.local_value, args)
```

**Review by fegin — APPROVED** [review]
LGTM, left some nits.

File: torch/distributed/_spmd/distribute.py (hunk 4 above)
Comment (mrshenli) [reply torch/distributed/_spmd/distribute.py:754]: Thanks! Will fix in a PR on top

**Pull request closed without merging**